### PR TITLE
Add constant propagation through methods with parameters or complex bodies

### PR DESCRIPTION
**Pull request by marek-safar**
This change also
- Adds simple inlining of method return values.
- Fixes several shortcomings of existing implementation like ignoring
methods called from unreachable blocks.

This helps to reduce the size of all runtime assemblies. For example size
of SPC for the default iOS project is reduced by 4% mostly due to metadata
removal. The linker speed improves a little bit (~3%) for the same
project.

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
I'm not actually reviewing this, but I had a couple areas of interest that I skimmed:

* Inlining string literals across modules is observable; we avoid doing it in CoreCLR: https://github.com/dotnet/runtime/blob/017744628361a6fc45c05c75cf640292c670a8dd/src/coreclr/vm/jitinterface.cpp#L8082-L8090
* Doing string concat ahead of time is observable too for similar reasons (if the newly concatenated string matches an existing literal). As the CoreCLR comment says "bugs are going to be tough to find". An example that has been brought up in the past is if someone takes a `lock` with such string, or uses `object.ReferenceEqual` because they expected a fresh `string`.
* Is there anything to prevent removing calls to methods on types with static constructors/module constructors?
* I think this is currently removing side effect for a callvirt with a null this (`static void Call(Foo instance) => instance.X();` - I saw a check for ldarg_0, but I didn't see a check for whether the calling method is instance).

**Comment by marek-safar** [comment]
> Inlining string literals across modules is observable; we avoid doing it in CoreCLR

Do you have an example why? I didn't understand the comment.

> Doing string concat ahead of time is observable too for similar reasons

Ugh, this is quite corner case but fair enough will address that.

> Is there anything to prevent removing calls to methods on types with static constructors/module constructors?

I had logic there but removed, now thinking about it again it needs more work even for other scenarios.

> I think this is currently removing side effect for a callvirt with a null this

It does not but there is a bug in the logic (missing a special case for static context).

**Comment by MichalStrehovsky** [comment]
>> Inlining string literals across modules is observable; we avoid doing it in CoreCLR

> Do you have an example why? I didn't understand the comment.

It has to do with CompilationRelaxations.NoStringInterning (that the C# compiler automatically emits into all assemblies it produces). The docs on it are pretty poor, here's StackOverflow: https://stackoverflow.com/questions/15601916/how-does-compilationrelaxations-nostringinterning-actually-work. That said, I looked into the implementation on the runtime side and at this point (after summer of last year) it seems like we no longer need to worry about it.

I created https://github.com/dotnet/runtime/pull/64521 to really double check it's no longer a concern. I expect JanK to scream if it's not the case and I'm wrong.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (jkotas) [inline src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:303]: This can create a very long string if called recursively.

Just curious - what are examples of CoreLib methods where this helps with trimming?

Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:303]: It does not matter if it creates long string as this logic is used for result computation only (value is not inlined). I found it for example here https://cs.github.com/dotnet/runtime/blob/f85ea976f81945ea18cd5dc71959cccecdc93cd2/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs?q=GetUtcFullDisplayName#L2111

**Comment by marek-safar** [comment]
> Is there anything to prevent removing calls to methods on types with static constructors/module constructors?

There is extra logic to take care of that.

> I think this is currently removing side effect for a callvirt with a null this (static void Call(Foo instance) => instance.X(); - I saw a check for ldarg_0, but I didn't see a check for whether the calling method is instance).

Yep, made the check more comprehensive.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (vitek-karas) [inline src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:1397]: This is very similar to MethodBodyScanner - might be worth looking into potential sharing. Or at the very least comparing the two implementations.

Comment (vitek-karas) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:1397]: That said - with the linker switching over to the "solver based data flow", the method body scanner will probably change quite a bit - especially around branching and such (basically it will become BasicBlockScanner instead).

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (MichalStrehovsky) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:460]: > @MichalStrehovsky might remember more details...

ConstantProp should be blocked for NoInlining methods because they specifically said they don't want to be inlined and it was breaking logic within library build (a constant method that we had a substitution for was getting inlined during library build, breaking the ability to substitute it at app build time).

Isn't constant prop blocked for NoInlining methods because of the check for `if (method.IsIntrinsic () || method.NoInlining)` above anyway? (Would we be able to get a MethodResult for it if we let it past this check here?)

Comment (vitek-karas) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:460]: You're right - I overlooked it in the existing code - please ignore this comment.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:454]: I think it was unintentional and probably not tested at all. I'd prefer to leave "static virtual" support to separate PR

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:518]: The purpose is actually different and the different behaviour is expected/wanted as inlining is much more restrictive than evaluation.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:499]: It can be both, shallow copy for complex instruction, deep copy for simple instruction. I have upstream PR which I'll merge here once approved (including the name)

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:464]: Added the deep stack check, the implementation goes only up to 3 iterations down on the default iOS app so I think this is fine.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #removed

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #fixed

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:305]: #done

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #done

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #fixed

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #fixed

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: yeah, not sure why though

#fixed

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #fixed the comment

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:429]: look ugly, kept it that way

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (vitek-karas) [inline src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:1418]: This is a weird way to handle backward branches. Maybe add a comment that this will effectively `break` on a backward jump.

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: #resolved

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs]: tweaked the comment

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:515]: comment added

File: src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs
Comment (marek-safar) [reply src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:1415]: commend added

**Review by vitek-karas — APPROVED** [review]
I think this looks good now.

**Pull request merged**
